SVPAlib/src/automata/sra/SRA.java

```diff
@@ -0,0 +1,1828 @@
+/**
+ * SVPAlib
+ * automata.sra
+ * Jul 25, 2018
+ * @author Tiago Ferreira
+ */
+package automata.sra;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+import org.sat4j.specs.TimeoutException;
+
+import automata.Automaton;
+import automata.Move;
+import theory.BooleanAlgebra;
+import utilities.Block;
+import utilities.Pair;
+import utilities.Timers;
+import utilities.UnionFindHopKarp;
+
+/**
+ * Symbolic Register Automaton
+ * 
+ * @param <P> set of predicates over the domain S
+ * @param <S> domain of the automaton alphabet
+ */
+public class SRA<P, S> extends Automaton<P, S> {
+	// ------------------------------------------------------
+	// Constant automata
+	// ------------------------------------------------------
+
+	public void setIsDet(boolean b) {
+		isDeterministic = b;
+	}
+
+	/**
+	 * Returns the empty SRA for the Boolean algebra <code>ba</code>
+	 * @throws TimeoutException 
+	 */
+	public static <A, B> SRA<A, B> getEmptySRA(BooleanAlgebra<A, B> ba) throws TimeoutException {
+		SRA<A, B> aut = new SRA<A, B>();
+		aut.states = new HashSet<Integer>();
+		aut.states.add(0);
+		aut.finalStates = new HashSet<Integer>();
+		aut.initialState = 0;
+        aut.registers = new HashSet<A>();
+        aut.registers.add(ba.False());
+		aut.isDeterministic = true;
+		aut.isEmpty = true;
+		aut.maxStateId = 1;
+        // FIXME: We require a register operation on every transition.
+        //        How are we handling an emptySRA? 
+		aut.addTransition(new SRACheckMove<A, B>(0, 0, ba.True(), 0), ba, true);
+		return aut;
+	}
+
+	/**
+	 * Returns the SRA accepting every string in the Boolean algebra
+	 * <code>ba</code>
+	 * @throws TimeoutException 
+	 */
+	public static <A, B> SRA<A, B> getFullSRA(BooleanAlgebra<A, B> ba) throws TimeoutException {
+		SRA<A, B> aut = new SRA<A, B>();
+		aut.states = new HashSet<Integer>();
+		aut.states.add(0);
+		aut.finalStates = new HashSet<Integer>(aut.states);
+		aut.initialState = 0;
+        aut.registers = new HashSet<A>();
+        aut.registers.add(ba.True());
+		aut.isDeterministic = true;
+		aut.isEmpty = false;
+		aut.maxStateId = 1;
+        // FIXME: We require a register operation on every transition.
```

Comment:
For the full SRA I added the register being a `ba.True()` so that no matter the input, it's always present in the register.

Reply:
I'm confused? Registers should store actual values not predicates. Also, you should allow more than 1 registers

Reply:
So Matteo had the idea for having the registers store predicates, more like a range of values really. This means they can also store single values as Atoms. Further discussion is definitely needed.

Reply:
We do allow for more than one register, right now registers are represented as a list. Each element being a different register, and the `Integer register` passed to the `Move`s is actually the index of the register in the list.

Reply:
We will indeed need to record predicates in the decision procedure but not in the model itself I believe. We should discuss. Good for the register(s)